src/Main.hs

```diff
@@ -80,195 +88,185 @@ markdownRouteFileBase :: MarkdownRoute -> Text
-
-  decodeRoute _model fp = do
-    if "static/" `T.isPrefixOf` toText fp
-      then pure $ Left fp
-      else do
-        if null fp
-          then pure $ Right indexMarkdownRoute
-          else do
-            basePath <- T.stripSuffix ".html" (toText fp)
-            slugs <- nonEmpty $ fromString . toString <$> T.splitOn "/" basePath
-            pure $ Right $ MarkdownRoute slugs
-
-  -- Routes to write when generating the static site.
-  allRoutes (Map.keys . modelDocs -> mdRoutes) =
-    [Left "static"]
-      <> fmap Right mdRoutes
+modelInsertStaticFile :: FilePath -> Model -> Model
+modelInsertStaticFile fp model = model {modelFiles = Set.insert fp (modelFiles model)}
+
+modelDeleteStaticFile :: FilePath -> Model -> Model
+modelDeleteStaticFile fp model = model {modelFiles = Set.delete fp (modelFiles model)}
 
 -- ------------------------
--- Main entry point
+-- Route encoder
 -- ------------------------
 
-log :: MonadLogger m => Text -> m ()
-log = logInfoNS "ema-template"
+newtype StaticRoute = StaticRoute {unStaticRoute :: FilePath}
+  deriving newtype (Eq, Ord, Show)
+
+-- A route encoder is simply a Prism with (some) model as its context.
+-- `mkRouteEncoder` enables you to create route encoders "manually" this way.
+instance IsRoute StaticRoute where
+  type RouteModel StaticRoute = Set FilePath
+  routeEncoder = mkRouteEncoder $ \files ->
+    let enc =
+          unStaticRoute
+        dec fp =
+          StaticRoute fp <$ guard (Set.member fp files)
+     in prism' enc dec
+  allRoutes files =
+    StaticRoute <$> toList files
+
+instance IsRoute MarkdownRoute where
+  type RouteModel MarkdownRoute = Set MarkdownRoute
+  routeEncoder = mkRouteEncoder $ \mdRoutes ->
+    let enc (MarkdownRoute slugs) =
+          toString $ T.intercalate "/" (Slug.unSlug <$> toList slugs)
+        dec fp = do
+          guard $ not $ null fp
+          slugs <- nonEmpty $ fromString . toString <$> T.splitOn "/" (toText fp)
+          let r = MarkdownRoute slugs
+          guard $ Set.member r mdRoutes
+          pure r
+     in htmlSuffixPrism % prism' enc dec
+  allRoutes =
+    toList
+
+-- Route encoders can also be composed. Here, we use `combineRouteEncoder` to
+-- combine two route encoders.
+instance IsRoute Route where
+  type RouteModel Route = Model
+  routeEncoder =
+    combineRouteEncoder
+      routeDecomposition
+      decomposeModel
+      (routeEncoder @MarkdownRoute)
+      (routeEncoder @StaticRoute)
+  allRoutes model =
+    fmap Route_Markdown (allRoutes @MarkdownRoute $ fst . decomposeModel $ model)
+      <> fmap Route_Static (allRoutes @StaticRoute $ snd . decomposeModel $ model)
+
+decomposeModel :: Model -> (Set MarkdownRoute, Set FilePath)
+decomposeModel = Map.keysSet . modelDocs &&& modelFiles
+
+routeDecomposition :: Iso' (Either MarkdownRoute StaticRoute) Route
+routeDecomposition =
+  iso (either Route_Markdown Route_Static) $ \case
+    Route_Markdown r -> Left r
+    Route_Static r -> Right r
```

Comment:
Too annoying to have the user do this. Ema should be able to generically derive this. https://github.com/srid/ema/pull/81#issuecomment-1163226891

Reply:
resolved in [f76280e](https://github.com/srid/ema-template/pull/22/commits/f76280e5189457a1ead3f2855bd03e62ca8cb8fc)